File: gfg/Middle of the Linked List.cpp
//https://leetcode.com/problems/middle-of-the-linked-list/

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        
        ListNode *fast=head;
        ListNode *slow=head;
        
        while(fast!=NULL && fast->next!=NULL){
            fast=fast->next->next;
            slow=slow->next;
        }
        return slow;

        
    }
};

File: LeetCode/Middle of the Linked List.rb
# https://leetcode.com/problems/middle-of-the-linked-list/

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def middle_node(head)
    fast, slow = head, head

    while fast != nil && fast.next != nil
        slow = slow.next
        fast = fast.next.next
    end
    return slow
end